pkg/util/metrics_helper.go

```diff
@@ -0,0 +1,311 @@
+		}
+
+		if err != nil {
+			level.Warn(Logger).Log("msg", "failed to gather metrics from TSDB shipper", "user", userID, "err", err)
+			continue
+		}
+	}
+	return data
+}
+
+// AddGatheredDataForUser adds user-specific output of Gatherer.Gather method.
+// Gatherer.Gather specifies that there metric families are uniquely named, and we use that fact here.
+// If they are not, this method returns error.
+func (d MetricFamiliesPerUser) AddGatheredDataForUser(userID string, metrics []*dto.MetricFamily) error {
+	// Keeping map of metric name to its family makes it easier to do searches later.
+	perMetricName := map[string]*dto.MetricFamily{}
+
+	for _, m := range metrics {
+		name := m.GetName()
+		// these errors should never happen when passing Gatherer.Gather() output.
+		if name == "" {
+			return errors.New("empty name for metric family")
+		}
+		if perMetricName[name] != nil {
+			return fmt.Errorf("non-unique name for metric family: %q", name)
+		}
+
+		perMetricName[name] = m
+	}
+
+	d[userID] = perMetricName
+	return nil
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCounters(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string) {
+	result := float64(0)
+	for _, perMetric := range d {
+		result += sum(perMetric[counter], counterValue)
+	}
+
+	out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, result)
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCountersWithLabels(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string, labelNames ...string) {
+	type counterResult struct {
+		value       float64
+		labelValues []string
+	}
+
+	result := map[string]counterResult{}
+
+	for _, userMetrics := range d { // for each user
+		metricsPerLabelValue := getMetricsWithLabelNames(userMetrics[counter], labelNames)
+
+		for key, mlv := range metricsPerLabelValue {
+			for _, m := range mlv.metrics {
+				r := result[key]
+				if r.labelValues == nil {
+					r.labelValues = mlv.labelValues
+				}
+
+				r.value += m.GetCounter().GetValue()
+				result[key] = r
+			}
+		}
+	}
+
+	for _, cr := range result {
+		out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, cr.value, cr.labelValues...)
+	}
+}
+
+func (d MetricFamiliesPerUser) SendSumOfCountersPerUser(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string) {
+	for user, perMetric := range d {
+		v := sum(perMetric[counter], counterValue)
+
+		out <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, v, user)
+	}
+}
+
+func (d MetricFamiliesPerUser) SendSumOfGauges(out chan<- prometheus.Metric, desc *prometheus.Desc, gauge string) {
+	result := float64(0)
+	for _, perMetric := range d {
+		result += sum(perMetric[gauge], gaugeValue)
+	}
+	out <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, result)
+}
+
+func (d MetricFamiliesPerUser) SendSumOfGaugesWithLabels(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string, labelNames ...string) {
+	type gaugeResult struct {
```

Comment:
This content is equal to `SendSumOfCountersWithLabels()`, except for the different `type` name (`counterResult` vs `gaugeResult`). We may generalize it into a `sumOfCollectorsWithLabels()` and then call it from `SendSumOf(Counters|Gauge)WithLabels()` to get the sum.

Reply:
Good idea, I'll do this change.